homeassistant/components/climate/coolmaster.py

```diff
@@ -0,0 +1,186 @@
+"""
+CoolMasterNet platform that offers control of CoolMasteNet Climate Devices.
+
+For more details about this platform, please refer to the documentation
+https://www.home-assistant.io/components/climate.coolmaster/
+"""
+
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components.climate import (
+    PLATFORM_SCHEMA, STATE_AUTO, STATE_COOL, STATE_DRY, STATE_FAN_ONLY,
+    STATE_HEAT, SUPPORT_FAN_MODE, SUPPORT_ON_OFF, SUPPORT_OPERATION_MODE,
+    SUPPORT_TARGET_TEMPERATURE, ClimateDevice)
+from homeassistant.const import (
+    ATTR_TEMPERATURE, CONF_HOST, CONF_PORT, TEMP_CELSIUS, TEMP_FAHRENHEIT)
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['pycoolmasternet==0.0.4']
+
+SUPPORT_FLAGS = (SUPPORT_TARGET_TEMPERATURE | SUPPORT_FAN_MODE |
+                 SUPPORT_OPERATION_MODE | SUPPORT_ON_OFF)
+
+DEFAULT_PORT = 10102
+
+AVAILABLE_MODES = [STATE_HEAT, STATE_COOL, STATE_AUTO, STATE_DRY,
+                   STATE_FAN_ONLY]
+
+CM_TO_HA_STATE = {
+    'heat': STATE_HEAT,
+    'cool': STATE_COOL,
+    'auto': STATE_AUTO,
+    'dry': STATE_DRY,
+    'fan': STATE_FAN_ONLY,
+}
+
+HA_STATE_TO_CM = {value: key for key, value in CM_TO_HA_STATE.items()}
+
+FAN_MODES = ['low', 'med', 'high', 'auto']
+
+CONF_SUPPORTED_MODES = 'supported_modes'
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+    vol.Optional(CONF_SUPPORTED_MODES, default=AVAILABLE_MODES):
+        vol.All(cv.ensure_list, [vol.In(AVAILABLE_MODES)]),
+})
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the CoolMasterNet climate platform."""
+    from pycoolmasternet import CoolMasterNet
+
+    supported_modes = config.get(CONF_SUPPORTED_MODES)
+    host = config[CONF_HOST]
+    port = config[CONF_PORT]
+    cool = CoolMasterNet(host, port=port)
+    devices = cool.devices()
+
+    for device in devices:
+        _LOGGER.debug("Found device %s", device.uid)
+
+    all_devices = [CoolmasterClimate(device, supported_modes)
+                   for device in devices]
```

Comment:
Append the created entities to the list in the first loop after the debug log instead. We'll save one loop.

Reply:
Not sure I agree with this. Since the number of items is capped by about 200, there's not going to be any major performance improvement here, and the runtime is anyway going to be dominated by the additional loop that gets the status of the devices after they're initialized.
I find that loops that combine a map and a side effect are usually more verbose, harder to read and more error prone than two separate loops.
Is there some Python syntax I don't know that would make this better? (this is my first real piece of code with Python, so my experience is based on other languages).

Reply:
As a compromise, I separated building the entity to its own function. Now there's only one loop, but the function has a side effect.